components/script/dom/imagedata.rs

```diff
@@ -27,31 +28,45 @@ pub struct ImageData {
 
 impl ImageData {
     #[allow(unsafe_code)]
-    pub fn new(global: &GlobalScope, width: u32, height: u32, mut data: Option<Vec<u8>>) -> Root<ImageData> {
-        let imagedata = box ImageData {
-            reflector_: Reflector::new(),
-            width: width,
-            height: height,
-            data: Heap::default(),
-        };
-        let len = width * height * 4;
-
+    pub fn from_vec(global: &GlobalScope, width: u32, height: u32, mut data: Option<Vec<u8>>) -> Root<ImageData> {
         unsafe {
+            let len = width * height * 4;
             let cx = global.get_cx();
-            rooted!(in (cx) let mut js_object = ptr::null_mut());
+            rooted!(in (cx) let mut typedarray = ptr::null_mut());
             let data = match data {
                 Some(ref mut d) => {
                     d.resize(len as usize, 0);
                     CreateWith::Slice(&d[..])
                 },
                 None => CreateWith::Length(len),
             };
-            Uint8ClampedArray::create(cx, data, js_object.handle_mut()).unwrap();
-            (*imagedata).data.set(js_object.get());
+            Uint8ClampedArray::create(cx, data, typedarray.handle_mut()).unwrap();
+            Self::new(global, width, height, Some(typedarray.get()))
         }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn new(global: &GlobalScope, width: u32, height: u32, data: Option<*mut JSObject>) -> Root<ImageData> {
```

Comment:
The whole method should be unsafe, and please rename with to `new_with_jsobject`.